komga/src/main/kotlin/org/gotson/komga/domain/service/LibraryScanner.kt

```diff
@@ -18,95 +26,238 @@ class LibraryScanner(
+      restoreDeletedSeries(scannedSeries)
+      updateSeriesAndBooks(scannedSeries, library)
+      restoreDeletedBooksAndAddNew(scannedSeries, library)
+    }.also { logger.info { "Library updated in $it" } }
+  }
+
+  private fun getBookHash(book: Book): String {
+    return bookRepository.findByLibraryIdAndUrlIncludeDeleted(book.libraryId, book.url)?.let {
+      if (isModified(it.fileLastModified, book.fileLastModified)) fileHasher.getHash(book.path())
+      else it.fileHash
+    } ?: fileHasher.getHash(book.path())
+  }
+
+  private fun deleteNoLongerExistingSeries(scannedSeries: Map<Series, Collection<Book>>, library: Library) {
+    if (scannedSeries.isEmpty()) {
+      logger.info { "Scan returned no series, mark all existing series as deleted" }
+      seriesRepository.findByLibraryId(library.id)
+        .map { it.id }
+        .let {
+          bookRepository.softDeleteByBookIds(bookRepository.findAllIdBySeriesIds(it))
+          seriesRepository.softDeleteAll(it)
+        }
+    } else {
+      scannedSeries.keys.map { it.url }.let { urls ->
+        val series = seriesRepository.findByLibraryIdAndUrlNotIn(library.id, urls).filterNot { it.deleted }
+        if (series.isNotEmpty()) {
+          logger.info { "Series is not on disk anymore, marking as deleted: $series" }
+          series.map { it.id }.let {
+            bookRepository.softDeleteByBookIds(bookRepository.findAllIdBySeriesIds(it))
+            seriesRepository.softDeleteAll(it)
           }
         }
       }
+    }
+  }
 
-      scannedSeries.forEach { (newSeries, newBooks) ->
-        val existingSeries = seriesRepository.findByLibraryIdAndUrl(library.id, newSeries.url)
-
-        // if series does not exist, save it
-        if (existingSeries == null) {
+  private fun updateSeriesAndBooks(scannedSeries: Map<Series, Collection<Book>>, library: Library) {
+    scannedSeries.forEach { (newSeries, newBooks) ->
+      seriesRepository.findByLibraryIdAndUrlIncludeDeleted(library.id, newSeries.url)?.let { existingSeries ->
+        if (existingSeries.deleted) {
+          seriesLifecycle.deleteOne(existingSeries.id)
           logger.info { "Adding new series: $newSeries" }
-          val createdSeries = seriesLifecycle.createSeries(newSeries)
-          seriesLifecycle.addBooks(createdSeries, newBooks)
-          seriesLifecycle.sortBooks(createdSeries)
-        } else {
-          // if series already exists, update it
-          logger.debug { "Scanned series already exists. Scanned: $newSeries, Existing: $existingSeries" }
-          val seriesChanged = newSeries.fileLastModified.truncatedTo(ChronoUnit.MILLIS) != existingSeries.fileLastModified.truncatedTo(ChronoUnit.MILLIS)
-          if (seriesChanged) {
-            logger.info { "Series changed on disk, updating: $existingSeries" }
-            seriesRepository.update(existingSeries.copy(fileLastModified = newSeries.fileLastModified))
+          seriesLifecycle.createSeries(newSeries)
```

Comment:
Not sure what's happening here. When a scanned Series matches a soft-deleted series, you delete the soft-deleted one, then recreate it?
That would also delete all soft-deleted books for good then, along with their metadata and read progress.